apps/studio/src/features/editing-experience/components/CreateFolderModal/CreateFolderModal.tsx

```diff
@@ -0,0 +1,173 @@
+import type { UseDisclosureReturn } from "@chakra-ui/react"
+import { useEffect } from "react"
+import {
+  Box,
+  FormControl,
+  FormHelperText,
+  FormLabel,
+  Icon,
+  Input,
+  Modal,
+  ModalBody,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+  VStack,
+} from "@chakra-ui/react"
+import {
+  Button,
+  FormErrorMessage,
+  ModalCloseButton,
+  useToast,
+} from "@opengovsg/design-system-react"
+import { BiLink } from "react-icons/bi"
+import { z } from "zod"
+
+import { useZodForm } from "~/lib/form"
+import {
+  createFolderSchema,
+  MAX_FOLDER_PERMALINK_LENGTH,
+  MAX_FOLDER_TITLE_LENGTH,
+} from "~/schemas/folder"
+import { trpc } from "~/utils/trpc"
+import { generateResourceUrl } from "../utils"
+
+type CreateFolderProps = z.infer<typeof createFolderSchema>
+
+type CreateFolderModalProps = Pick<UseDisclosureReturn, "isOpen" | "onClose"> &
+  Pick<CreateFolderProps, "siteId"> &
+  Pick<CreateFolderProps, "parentFolderId">
+
+export const CreateFolderModal = ({
+  isOpen,
+  onClose,
+  siteId,
+  parentFolderId,
+}: CreateFolderModalProps): JSX.Element => {
+  const { setValue, register, handleSubmit, watch, formState, getFieldState } =
+    useZodForm({
+      defaultValues: {
+        folderTitle: "",
+        permalink: "",
+      },
+      schema: createFolderSchema.omit({ siteId: true, parentFolderId: true }),
+    })
+  const { errors, isValid } = formState
+  const utils = trpc.useUtils()
+  const toast = useToast()
+  const { mutate, isLoading } = trpc.folder.create.useMutation({
+    onSettled: onClose,
+    onSuccess: () => {
+      void utils.site.list.invalidate()
+      void utils.page.list.invalidate()
+      toast({ title: "Folder created!", status: "success" })
+    },
+    onError: (err) => {
+      toast({
+        title: "Failed to create folder",
+        status: "error",
+        // TODO: check if this property is correct
+        description: err.message,
+      })
+    },
+  })
+
+  const onSubmit = handleSubmit((data) => {
+    mutate({ ...data, parentFolderId, siteId })
+  })
+
+  const [folderTitle, permalink] = watch(["folderTitle", "permalink"])
+
+  useEffect(() => {
+    const permalinkFieldState = getFieldState("permalink")
+    // This allows the syncing to happen only when the page title is not dirty
+    // Dirty means user has changed the value AND the value is not the same as the default value of "".
+    // Once the value has been cleared, dirty state will reset.
+    if (!permalinkFieldState.isDirty) {
+      setValue("permalink", generateResourceUrl(folderTitle), {
+        shouldValidate: !!folderTitle,
+      })
+    }
+  }, [getFieldState, setValue, folderTitle])
```

Comment:
there's an issue with this (and page creation) where the `title` doesn't show errors on bounds exceeded.

additionally, if you exceed `permalink.maxLength`, the error doesn't go away too (for both) - i suspect it's this hook but i'm living with it for now